be/src/olap/schema_change.cpp

```diff
@@ -958,6 +1232,270 @@ Status SchemaChangeHandler::_do_process_alter_tablet_v2(const TAlterTabletReqV2&
+
+        do {
+            RowsetSharedPtr max_rowset;
+            // get history data to rebuild inverted index and it will check if there is hold in base tablet
+            res = _get_versions_to_be_changed(tablet, &versions_to_be_changed, &max_rowset);
+            if (!res.ok()) {
+                LOG(WARNING) << "fail to get version to be rebuild inverted index. res=" << res;
+                break;
+            }
+
+            // should check the max_version >= request.alter_version, if not the rebuild index is useless
+            if (max_rowset == nullptr || max_rowset->end_version() < request.alter_version) {
+                LOG(WARNING) << "base tablet's max version="
+                             << (max_rowset == nullptr ? 0 : max_rowset->end_version())
+                             << " is less than request version=" << request.alter_version;
+                res = Status::InternalError(
+                        "base tablet's max version={} is less than request version={}",
+                        (max_rowset == nullptr ? 0 : max_rowset->end_version()),
+                        request.alter_version);
+                break;
+            }
+
+            // init one delete handler
+            int64_t end_version = -1;
+            for (auto& version : versions_to_be_changed) {
+                end_version = std::max(end_version, version.second);
+            }
+
+            auto& all_del_preds = tablet->delete_predicates();
+            for (auto& delete_pred : all_del_preds) {
+                if (delete_pred->version().first > end_version) {
+                    continue;
+                }
+                tablet_schema->merge_dropped_columns(tablet->tablet_schema(delete_pred->version()));
+            }
+            res = delete_handler->init(tablet_schema, all_del_preds, end_version);
+            if (!res) {
+                LOG(WARNING) << "init delete handler failed. tablet=" << tablet->full_name()
+                             << ", end_version=" << end_version;
+                break;
+            }
+
+            // acquire data sources correspond to history versions
+            tablet->capture_rs_readers(versions_to_be_changed, rs_readers);
+            if (rs_readers->size() < 1) {
+                LOG(WARNING) << "fail to acquire all data sources. "
+                             << "version_num=" << versions_to_be_changed.size()
+                             << ", data_source_num=" << rs_readers->size();
+                res = Status::Error<ALTER_DELTA_DOES_NOT_EXISTS>();
+                break;
+            }
+
+            // reader_context is stack variables, it's lifetime should keep the same with rs_readers
+            RowsetReaderContext reader_context;
+            reader_context.reader_type = READER_ALTER_TABLE;
+            reader_context.tablet_schema = tablet_schema;
+            reader_context.need_ordered_result = true;
+            reader_context.delete_handler = delete_handler;
```

Comment:
If delete_handler filter rows that's marked deleted, the index will get rows not exactly the same as segment data file, and then MATCH query will produce wrong result. This need to be tested carefully.

Reply:
I will add more regression test case